### DISCUSSION TITLE
How to get data in a time series classifiction compatible format; relation of directional forecasting to TSC

### DISCUSSION THREAD

**Discussion by ArnonB** [alan-turing-institute/sktime#2266] [Q&A]
Hi, I am new to sktime, and I would like to ask about basic time classification:
I have X as NumPy matrix, in the common format, where each row is another day sample, and each column describe a different feature (of one value in each cell). (the 2 features are financial data with no connections between them)
[](https://i.stack.imgur.com/XNhjC.jpg)I would like to do classification with sktime. I have X as NumPy matrix, in the common format, where each row is another day sample, and each column describe a different feature of one value. (the 2 features are financial data with no connections between them)

![image](https://user-images.githubusercontent.com/64684799/159226009-3e594989-b95e-4519-b4c4-6a0849b747ad.png)

I have 3 questions:

1) How to convert the two features (X) to the Dataframe format needed by sktime for classification? as described here: 
    (https://www.sktime.org/en/stable/examples/loading_data.html)

2) In such case like the above (financial data) does sktime have any advantage over using sklearn classifiers? (for example, does using 
    KNeighborsTimeSeriesClassifier of sktime will have any advantage over using sklearn knn (KNeighborsClassifier)

3) If yes, what sktime classes are most suitable for financial data classification of trends directions, when the financial data is consists 
    of simple features like in the example above.

**Thank you very much!!**

**Comment by fkiraly** [comment]
@ArnonB, from your description it looks like you have a single time series, can you kindly confirm?

Time series classification is attaching labels to time series, one label per individual time series, on a batch of multiple time series, after having trained the classifier on multiple pairs of time series and ground truth label.

However, you seem to have only a single time series, hence time series classification does not seem appropriate.

Or, do you want to do sth different from time series classification?

**Reply by ArnonB** [reply]
Hi Franz, thanks for your reply.
In my example, the time is advancing only from **row to row**. On each row there are financial features of that day, i.e. their value at that day only. (each column for another feature). 
(Beside that, for each day (row) I have a different array of categories 0, 1, 2.)
From your answer, I understand that sktime is not fitting for such classification, although time is involve here (from row to row).
And if I understand correctly, only if each row will consists of features where each one of them is time series for its own, 
then I can use sktime. 
Is it correct?

**Reply by fkiraly** [reply]
> only if each row will consists of features where each one of them is time series for its own,
then I can use sktime. Is it correct?

No, that's not entirely correct, as a sumamry of what I meant.
sktime supports other learning tasks as well, e.g., forecasting, but that task is not "time series classification".

If I may clarify more what I was trying to say:
* I haven't fully understood what your intended learning task is
* but from your description I'm reasonably certain it's not time series classification (as you mentioned)
* sktime supports time series classification, but as said that's probably not what you want, e.g., you probably don't want `KNeighborsTimeSeriesClassifier` etc
* sktime supports other learning tasks as well, hence may or may not support the task you want. But that I can answer only once I understand what it is you want.

Does that make sense 😃 ?
Sorry for not being clear.

**Reply by fkiraly** [reply]
a good starter question for diagnosing the task is as follows.

When you are trying to make a prediction in the intended final use case: which data do you know? Which data do you want to predict?

Perhaps you could describe the data known at time of prediction, and the entry you want to predict, in a couple cases, in an example?

**Reply by ArnonB** [reply]
I will give example to make it more clear:
For each day (row), I want to try predict if the euro/dollar exchange rate will decline (0), have no change (1), or go up (2)
at the end of next 7 days. (3 categories classification)
my features (X) for each day, are : euro/dollar exchange rate, gold price, amount of money, news ranking as integer , etc.
(X is NumPy 2-d matrix, where each feature is another column, and each row represent another day)
(So time 'advancing' from row to row only, and the classification itself also has connection to future date)
So time has general connection to all of it. Does sktime can help in such case, (the aforementioned classification), 
And also, may sktime be useful just to try to predict a future euro/dollar rate with the above features (i.e, regression problem)

**Reply by fkiraly** [reply]
Ah, thanks for the explanation, @ArnonB.

Ok, there are two tasks:

The *first* task you describe is called "directional forecasting", it is not supported by `sktime` currently.
There is an open issue on support for it: https://github.com/alan-turing-institute/sktime/issues/745

You can, however, reduce it to time series classification manually, by doing the following:
Use a sliding window approach to pair up a window of a time series, and the up/down/mid information, in reference to the last time point as a label. Then, apply a time series classifier from the TSC module.

The *second* task is simply called forecasting, with exogeneous features that are gold price, amount of money, etc.
This is well-covered by the forecasting module, see the tutorial on forecasting.
Go here to look at an editable/playable version of the tutorial notebook:
![image](https://user-images.githubusercontent.com/7985502/160304458-1c4b3d2c-988b-4ae0-952a-ce9fecff8d11.png)

**Reply by fkiraly** [reply]
To come back to your original question about the data format for TSC, there are multiple options.

Have a look at the notebook `AA_datatypes_and_datasets` (same link as for forecasting tutorial), you need to bring your dataset into one of the `Panel` formats.
Depending on what your original data is in, you may find one of the following formats easier than the others: `pd-multiindex`, `nested_univ`, `numpy3D`, or `df-list`. you can use the `check_raise` utility to see whether your data container complies with the chosen format.

**Reply by ArnonB** [reply]
Thank you very much for your reply and support, I appreciate it!

**Reply by fkiraly** [reply]
alright - if you feel satisfied with the answer, kindly click "mark as answer"

**Reply by ArnonB** [reply]
👍done
